cmd/provider/main.go

```diff
@@ -0,0 +1,172 @@
+/*
+Copyright 2021 The Crossplane Authors.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package main
+
+import (
+	"encoding/json"
+	"fmt"
+	"os"
+	"time"
+
+	"github.com/pkg/errors"
+	"github.com/spf13/cobra"
+	"gopkg.in/yaml.v2"
+
+	"github.com/crossplane/crossplane-runtime/pkg/logging"
+
+	"github.com/crossplane-contrib/terrajet/pkg/cmdconfig"
+	"github.com/crossplane-contrib/terrajet/pkg/log"
+	"github.com/crossplane-contrib/terrajet/pkg/process"
+	"github.com/crossplane-contrib/terrajet/pkg/version"
+)
+
+const (
+	// command-line flags for nats-monitoring executable
+	flagDebug      = "debug"
+	flagTimeout    = "timeout"
+	flagPort       = "port"
+	flagPluginPath = "plugin-path"
+	// default values for command-line arguments
+	valTimeout    = 30 * time.Second
+	valPort       = 9000
+	valPluginPath = "/usr/local/bin/provider"
+	// config file path env. variable
+	appConfigFileEnvVarName = "APP_CONFIG_PATH"
+	// provider plugin binary path
+)
+
+// ProviderConfig maintains the configurations for the
+// provider runner
+type ProviderConfig struct {
+	// Debug enables debug level logging
+	Debug bool `mapstructure:"debug" yaml:"debug"`
+	// Timeout for reading the address info from provider's stdout
+	Timeout time.Duration `mapstructure:"timeout" yaml:"timeout"`
+	// Port the provider will listen on
+	Port int `mapstructure:"port" yaml:"port"`
+	// PluginPath path of the provider executable
+	PluginPath string `mapstructure:"plugin-path" yaml:"plugin-path"`
+}
+
+func (c ProviderConfig) String() (string, error) {
+	buff, err := yaml.Marshal(c)
+	if err != nil {
+		return "", err
+	}
+	return string(buff), nil
+}
+
+var logger logging.Logger
+
+func main() {
+	conf := &ProviderConfig{}
+	// initialize a cobra command for nats-monitoring
```

Comment:
I don't think this is for nats-monitoring. :)

Reply:
Resolved via removal (please see the comment below on the removal of the provider wrapper implementation).

Reply:
Initially, I believe it makes sense to continue without the shared gRPC server and only with provider binary mirroring for production, so that we can safely invoke `terraform init`s in synchronous contexts and we are doing things in the documented ways Terraform expects us to do. Shared gRPC server is certainly a hack that relies on undocumented features used for testing providers and running the Terraform CLI on Windows.

But it looks like we have an option if we do hit certain scalability issues in the long run. The idea is still to run the Terraform CLI behind the scenes, but instead of letting the Terraform commands fork provider gRPC servers when they are about to make gRPC calls to the provider, we have (for now) an option to make them share a provider gRPC server.